topk panics when param < 1

```
		{
			name: "bbb",
			load: `load 30s
http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
http_requests_total{pod="nginx-2", series="2"} 2+2.3x50
http_requests_total{pod="nginx-3", series="3"} 6+0.8x60
http_requests_total{pod="nginx-4", series="3"} 5+2.4x50
http_requests_total{pod="nginx-5", series="1"} 8.4+2.3x50
http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50
`,
			query: `topk without () (0.5, vector(time()))`,
		}
```

Using float value for the `k` param is actually valid in the old Prometheus engine. Same query won't hit panic in the old engine but panics in the new engine.

Old engine: https://prometheus.demo.do.prometheus.io/graph?g0.expr=topk%20without%20()%20(0.5%2C%20vector(time()))&g0.tab=1&g0.stacked=0&g0.range_input=1h

```
=== RUN   TestInstantQuery/disableOptimizers=true/bbb/disableFallback=false
panic: runtime error: index out of range [0] with length 0

goroutine 108 [running]:
github.com/thanos-community/promql-engine/execution/aggregate.(*kAggregate).aggregate(0x14000546a90, 0xc350, 0x14000077de8, 0x0, {0x1400032ce30, 0x1, 0x1?}, {0x140007d00e0, 0x1, 0x0?})
	/Users/benye/hub/gowork/promql-engine/execution/aggregate/khashaggregate.go:167 +0x414
github.com/thanos-community/promql-engine/execution/aggregate.(*kAggregate).Next(0x14000546a90, {0x1012b3ec8, 0x14000115f80})
	/Users/benye/hub/gowork/promql-engine/execution/aggregate/khashaggregate.go:114 +0x594
github.com/thanos-community/promql-engine/execution/exchange.(*concurrencyOperator).pull(0x14000801a10, {0x1012b3ec8, 0x14000115f80})
	/Users/benye/hub/gowork/promql-engine/execution/exchange/concurrent.go:80 +0xc0
created by github.com/thanos-community/promql-engine/execution/exchange.(*concurrencyOperator).Next.func1
	/Users/benye/hub/gowork/promql-engine/execution/exchange/concurrent.go:56 +0xa0
```

A very similar issue related to this param https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L2334.
The old engine will check the whether the param is within the int64 range.

```
bottomk by (series, pod) ((time() ^ time()), {series="2"} offset -1m52s)
    compatibility_test.go:109: 
        	Error Trace:	/Users/benye/hub/gowork/promqlsmith/example/thanos-engine/compatibility_test.go:109
        	Error:      	Not equal: 
        	            	expected: *errors.errorString(&errors.errorString{s:"Scalar value +Inf overflows int64"})
        	            	actual  : <nil>(<nil>)
        	Test:       	TestPromQLCompatibility
--- FAIL: TestPromQLCompatibility (5.12s)
```

We don't have that check in the new engine so error mismatched.

--- a/Makefile
+++ b/Makefile
@@ -35,7 +35,7 @@ export GOCACHE=/tmp/cache
 test:
 	@echo ">> running unit tests (without cache)"
 	@rm -rf $(GOCACHE)
-	@go test -v -race -timeout=5m $(shell go list ./...);
+	@go test -v -race -timeout=10m $(shell go list ./...);
 
 .PHONY: deps
 deps: ## Ensures fresh go.mod and go.sum.

--- a/engine/engine_test.go
+++ b/engine/engine_test.go
@@ -1462,6 +1462,74 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
 			step:  2 * time.Second,
 		},
 		{
+			name: "topk with float64 parameter",
+			load: `load 30s
+				http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
+				http_requests_total{pod="nginx-2", series="1"} 2+2.3x50
+				http_requests_total{pod="nginx-4", series="2"} 5+2.4x50
+				http_requests_total{pod="nginx-5", series="2"} 8.4+2.3x50
+				http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50`,
+			query: "topk(3.5, http_requests_total)",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
+			name: "topk with float64 parameter that gets truncated to 0",
+			load: `load 30s
+				http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
+				http_requests_total{pod="nginx-2", series="1"} 2+2.3x50
+				http_requests_total{pod="nginx-4", series="2"} 5+2.4x50
+				http_requests_total{pod="nginx-5", series="2"} 8.4+2.3x50
+				http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50`,
+			query: "topk(0.5, http_requests_total)",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
+			name: "topk with float64 parameter that does not fit int64",
+			load: `load 30s
+				http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
+				http_requests_total{pod="nginx-2", series="1"} 2+2.3x50
+				http_requests_total{pod="nginx-4", series="2"} 5+2.4x50
+				http_requests_total{pod="nginx-5", series="2"} 8.4+2.3x50
+				http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50`,
+			query: "topk(1e120, http_requests_total)",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
+			name: "topk with NaN",
+			load: `load 30s
+				http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
+				http_requests_total{pod="nginx-2", series="1"} 2+2.3x50
+				http_requests_total{pod="nginx-4", series="2"} 5+2.4x50
+				http_requests_total{pod="nginx-5", series="2"} 8.4+2.3x50
+				http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50`,
+			query: "topk(NaN, http_requests_total)",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
+			name:  "topk with NaN and no matching series",
+			query: "topk(NaN, not_there)",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
+			name: "nested topk error that should not be skipped",
+			load: `load 30s
+				X 1+1x50`,
+			query: "topk (0, topk (NaN, X))",
+			start: time.Unix(0, 0),
+			end:   time.Unix(3000, 0),
+			step:  2 * time.Second,
+		},
+		{
 			name: "topk wrapped by another aggregate",
 			load: `load 30s
 				http_requests_total{pod="nginx-1", series="1"} 1+1.1x40

--- a/execution/aggregate/khashaggregate.go
+++ b/execution/aggregate/khashaggregate.go
@@ -82,25 +82,28 @@ func (a *kAggregate) Next(ctx context.Context) ([]model.StepVector, error) {
 	if err != nil {
 		return nil, err
 	}
-	if in == nil {
-		return nil, nil
-	}
-
-	defer a.next.GetPool().PutVectors(in)
-
 	args, err := a.paramOp.Next(ctx)
 	if err != nil {
 		return nil, err
 	}
-	for i := range a.params {
-		a.params[i] = math.NaN()
-		if i < len(args) {
-			a.params[i] = args[i].Samples[0]
-			a.paramOp.GetPool().PutStepVector(args[i])
+
+	for i := range args {
+		a.params[i] = args[i].Samples[0]
+		a.paramOp.GetPool().PutStepVector(args[i])
+
+		val := a.params[i]
+		if val > math.MaxInt64 || val < math.MinInt64 || math.IsNaN(val) {
+			return nil, errors.Newf("scalar value %v overflows int64", val)
+		}
+		if int(val) == 0 {
+			return nil, nil
 		}
 	}
 	a.paramOp.GetPool().PutVectors(args)
 
+	if in == nil {
+		return nil, nil
+	}
 	if len(args) < len(in) {
 		return nil, errors.New("scalar argument not found")
 	}
@@ -115,6 +118,7 @@ func (a *kAggregate) Next(ctx context.Context) ([]model.StepVector, error) {
 		a.aggregate(vector.T, &result, int(a.params[i]), vector.SampleIDs, vector.Samples)
 		a.next.GetPool().PutStepVector(vector)
 	}
+	a.next.GetPool().PutVectors(in)
 
 	return result, nil
 }

